File: Between/Between/Services/EventsBuilder.swift
import Foundation

enum EventsBuilder {
    static func build(
        events: [CampusEvent],
        interests: [Interest],
        participations: [EventParticipation],
        partnerProfiles: [PartnerSeekingProfile],
        students: [Student],
        viewerId: String,
        myInterestIds: [String],
        activeMode: ActivityMode?,
        modeExpiresAt: Date?,
        onboardingComplete: Bool
    ) -> EventsData {
        let interestById = Dictionary(uniqueKeysWithValues: interests.map { ($0.id, $0) })
        let namesById = Dictionary(uniqueKeysWithValues: students.map { ($0.id, $0.name) })

        let viewerParticipations = participations.filter { $0.studentId == viewerId }
        let viewerPartnerEventIds = Set(viewerParticipations.filter { $0.kind == .lookingForPartner }.map(\.eventId))

        let cards: [CampusEventCard] = events
            .sorted { $0.startTime < $1.startTime }
            .map { event in
                let eventParts = participations.filter { $0.eventId == event.id }
                let realInterested = Set(eventParts.map(\.studentId)).count
                let realPartners = eventParts.filter { $0.kind == .lookingForPartner }.count
                let interestedCount = realInterested + event.promotedInterestedCount
                let partnerCount = realPartners + event.promotedPartnerCount
                let isInterested = eventParts.contains { $0.studentId == viewerId }
                let isLooking = viewerPartnerEventIds.contains(event.id)
                let canView = isLooking && event.matchingKind != .none

                let profiles: [PartnerSeekingProfile] = canView
                    ? partnerProfiles.filter { $0.eventId == event.id && $0.studentId != viewerId }
                    : []

                let interest = interestById[event.interestId]
                let end = event.endTime ?? event.startTime.addingTimeInterval(3600)
                let timeLabel = ScheduleEngine.formatRange(
                    start: minutes(from: event.startTime),
                    end: minutes(from: end)
                )

                return CampusEventCard(
                    id: event.id,
                    title: event.title,
                    description: event.description,
                    location: event.location,
                    timeLabel: timeLabel,
                    interestName: interest?.name ?? "Campus",
                    interestIcon: interest?.icon ?? "calendar",
                    interestedCount: interestedCount,
                    partnerSeekingCount: partnerCount,
                    matchingKind: event.matchingKind,
                    recurrenceLabel: event.recurrenceLabel,
                    isInterested: isInterested,
                    isLookingForPartner: isLooking,
                    canViewPartners: canView,
                    partnerProfiles: profiles
                )
            }

        let sorted = cards.sorted { lhs, rhs in
            let lhsMatch = myInterestIds.contains(events.first(where: { $0.id == lhs.id })?.interestId ?? "")
            let rhsMatch = myInterestIds.contains(events.first(where: { $0.id == rhs.id })?.interestId ?? "")
            if lhsMatch != rhsMatch { return lhsMatch }
            return lhs.interestedCount > rhs.interestedCount
        }

        return EventsData(
            events: sorted,
            interests: interests,
            myInterestIds: myInterestIds,
            onboardingComplete: onboardingComplete,
            activeMode: activeMode,
            modeExpiresAt: modeExpiresAt
        )
    }

    private static func minutes(from date: Date) -> Int {
        let cal = Calendar.current
        return cal.component(.hour, from: date) * 60 + cal.component(.minute, from: date)
    }
}

File: Between/api/v1/eventsBuilder.js
const { formatRange } = require('./scheduleEngine');

function buildEvents(state, viewerId) {
  const profile = state.studentProfiles.find((p) => p.studentId === viewerId);
  const modeEntry = state.activeModeByStudentId[viewerId];
  const student = state.students.find((s) => s.id === viewerId);
  const schoolId = student?.schoolId || 'vt';

  const events = state.campusEvents.filter((e) => e.schoolId === schoolId);
  const interests = state.interests.filter((i) => i.schoolId === schoolId);
  const myInterestIds = profile?.interestIds || [];

  const viewerPartnerEvents = new Set(
    state.eventParticipations
      .filter((p) => p.studentId === viewerId && p.kind === 'lookingForPartner')
      .map((p) => p.eventId)
  );

  const cards = events.map((event) => {
    const parts = state.eventParticipations.filter((p) => p.eventId === event.id);
    const realInterested = new Set(parts.map((p) => p.studentId)).size;
    const realPartners = parts.filter((p) => p.kind === 'lookingForPartner').length;
    const interestedCount = realInterested + (event.promotedInterestedCount || 0);
    const partnerSeekingCount = realPartners + (event.promotedPartnerCount || 0);
    const isInterested = parts.some((p) => p.studentId === viewerId);
    const isLooking = viewerPartnerEvents.has(event.id);
    const matchingKind = event.matchingKind || 'partner';
    const canView = isLooking && matchingKind !== 'none';

    const interest = interests.find((i) => i.id === event.interestId);
    const end = event.endTime ? new Date(event.endTime) : new Date(new Date(event.startTime).getTime() + 3600000);
    const startM = minutesFromDate(new Date(event.startTime));
    const endM = minutesFromDate(end);

    const partnerProfiles = canView
      ? state.partnerProfiles.filter((p) => p.eventId === event.id && p.studentId !== viewerId)
      : [];

    return {
      id: event.id,
      title: event.title,
      description: event.description,
      location: event.location,
      timeLabel: formatRange(startM, endM),
      interestName: interest?.name || 'Campus',
      interestIcon: interest?.icon || 'calendar',
      interestedCount,
      partnerSeekingCount,
      matchingKind,
      recurrenceLabel: event.recurrenceLabel || null,
      isInterested,
      isLookingForPartner: isLooking,
      canViewPartners: canView,
      partnerProfiles,
    };
  });

  cards.sort((a, b) => {
    const aMatch = interestMatch(a, events, myInterestIds);
    const bMatch = interestMatch(b, events, myInterestIds);
    if (aMatch !== bMatch) return aMatch ? -1 : 1;
    return b.interestedCount - a.interestedCount;
  });

  return {
    events: cards,
    interests,
    myInterestIds,
    onboardingComplete: profile?.onboardingComplete ?? false,
    activeMode: modeEntry?.mode || null,
    modeExpiresAt: modeEntry?.expiresAt || null,
  };
}

function interestMatch(card, events, myInterestIds) {
  const ev = events.find((e) => e.id === card.id);
  return ev ? myInterestIds.includes(ev.interestId) : false;
}

function minutesFromDate(d) {
  return d.getHours() * 60 + d.getMinutes();
}

module.exports = { buildEvents };
